python/zygote.py

```diff
@@ -74,166 +75,177 @@ def try_dumps(obj, sort_keys=False, allow_nan=False):
         raise
 
 
-def worker_loop():
+def worker_loop(s: socket.socket):
```

Comment:
A good opportunity to add type hints to this whole file. Don't forget to add zygote.py to the pyproject file.

Reply:
`asyncio` only really helps if you're trying to do multiple things in parallel, right? All this code is intentionally very sequential, so I'm not sure what improvements we'd expect to see.

Reply:
It's helpful in situations where you have blocking calls (like writing or reading data to a socket here). If you want to keep the changes to this file minimal, it's probably not worth pursuing, but the framework has builtin constructs for an event loop and signal handling that simplify some of the other code in this file https://docs.python.org/3/library/asyncio-stream.html#register-an-open-socket-to-wait-for-data-using-streams

Reply:
But does it actually change anything if there's not other work that could be done while waiting for those operations to complete? i.e., if I was writing a Node server that only handled one request at a time, it wouldn't matter if I used sync vs async filesystem operations, as I already wouldn't be using the main thread for other things while waiting for those FS operations to complete. Does `async`/`await` somehow make async operations faster in Python? I wouldn't expect that to be the case.

Reply:
Does PL expect requests sent to this file to be served in the order they were received? If that's the case then I think you're right that there's no real way to speed this up. `async`/`await` in Python basically just allows you to have your code do other work while waiting for blocking calls to finish (i.e. serialize JSON while waiting to read something from the file system). But if you want to force requests to be handled in-order there's not much that can be done.

Reply:
PL only sends a new request once the previous one is done; there will never be multiple pending or in-flight IPCs at the time time. This is very much intentional: elements + question code _must_ be processed sequentially (since they mutate the shared `data` object), and we can't have a single Python worker doing work for multiple questions at once because we only a expose a single course directory to the worker at a time to prevent course A from accessing course B's files.

Reply:
Gotcha, makes sense. Yeah then the main benefit of using async is not having to manually fork the process and cleaner handling of the event loop. Something to consider if this file gets reworked but to minimize changes then shouldn't be included here.

Reply:
Manual forking is still necessary for isolation, right?

Reply:
I think so, but I'm not even sure the zygote relies on this? The main benefit would be simplifying the `worker_loop` function, the external fork creation can still happen.